Explore: Allow the use of plugin panels

Umaaz opened the pull request:
Allow plugins to define a visualisation to use in explore that comes from a plugin.

<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

This PR allows a datasource plugin to specify a panel from a panel plugin to use when rendering the query result on the explore page.

By setting `frame.meta.preferredVisualisationType` to `plugin` and adding `frame.meta.custom.pluginID` to the dataframe response from the datasource. The explore view can load the appropriate panel.

**Why do we need this feature?**

I needed this feature, as I am developing a datasource that stores data that doesn't fall into the categories provided by the current explore page.

**Who is this feature for?**

Users that are developing datasources that provided data that requires specific rendering, and want to take advantage of the explore functionality.

**Which issue(s) does this PR fix?**:

I could not find any specific issues that relate to this, possibly relates to #58849, but not sure.

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

Umaaz committed:
Explore: Allow the use of plugin panels

Allow plugins to define a visualisation to use in explore that comes from a plugin.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=66982) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=66982) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=66982) it.</sub>

academo commented:
@Umaaz can you provide a code example of how this looks from the plugin's code perspective?

marefr commented on public/app/features/explore/utils/decorators.ts:
```diff
@@ -58,6 +59,11 @@ export const decorateWithFrameTypeMetadata = (data: PanelData): ExplorePanelData
       case 'flamegraph':
         flameGraphFrames.push(frame);
         break;
+      case 'plugin':
```
Looks like this potentially could be part of the default case and if set, try using the plugin id and load the plugin. Then preferredVisualization type could support any plugin id, rather than introducing a type=plugin in the mix? Guess it might complicate things?

Umaaz replied:
I did do that originally, then thought it would be better to have a explicit type=plugin to prevent any chance of existing plugins causing issues. In the very small chance that some plugin exists with `meta.custom.pluginID`.

gelicia commented:
We need to discuss this a little more internally but so far I'm liking it. 

Posting testing instructions if anyone else wants to play around.

Tested with the following:
Test Datasource, raw frames scenario
The following raw dataframe:
<details>
<summary>Dataframe</summary>
<br>
```
[
  {
    "meta": {
      "preferredVisualisationType": "plugin",
      "custom": {
        "pluginID": "piechart"
      }
    },
    "fields": [
      {
        "name": "time",
        "values": [
          1,
          2
        ]
      },
      {
        "name": "log",
        "values": [
          "app=foo level=error msg=\"Request\" endpoint=/test",
          "app=foo level=info msg=\"Deployed\""
        ]
      },
      {
        "name": "app",
        "values": [
          "foo",
          "foo"
        ]
      },
      {
        "name": "endpoint",
        "values": [
          "/test",
          null
        ]
      }
    ]
  }
]
```
</details>

Changed around the pluginID to 'table', 'logs', 'stat', 'cloudwatch', 'dashlist', etc
Explore did its best to render the data, throwing an error if it didn't work because the data wasn't formatted in a way the visualization would expect.

gelicia commented on public/app/features/explore/ExplorePanel.tsx:
```diff
@@ -0,0 +1,50 @@
+  state: LoadingState;
+}
+
+export function ExplorePanel({ width, height, timeZone, state, frame, absoluteRange }: Props) {
```
Wondering what you think about naming this something like `CustomPanel` so it doesn't sound so generic.

Umaaz replied:
Seems to be a bit of a mix, I will rename it to `CustomContainer` as it seem to fit best.

Umaaz replied:
I have now renamed the container.

torkelo commented:
if we need a new custom meta field it might be good to add frame.meta.preferredVisualisationPluginId ?

torkelo commented on public/app/features/explore/ExplorePanel.tsx:
```diff
@@ -0,0 +1,50 @@
+    <PanelChrome title={title} width={width} height={height} loadingState={state}>
+      {(innerWidth, innerHeight) => (
+        <PanelRenderer
+          data={{ series: [frame], state: state, timeRange }}
```
why is only a single frame passed to the visualization?

marefr commented:
Reminder. If merging this we need to update https://github.com/grafana/grafana-plugin-sdk-go/blob/0c55ef8790683bbf1f4aa26e7edb625becad9e2d/data/frame_meta.go#L54-L72 

Given
> if we need a new custom meta field it might be good to add frame.meta.preferredVisualisationPluginId ?

We need to add it to https://github.com/grafana/grafana-plugin-sdk-go/blob/0c55ef8790683bbf1f4aa26e7edb625becad9e2d/data/frame_meta.go#L14

Umaaz committed:
Explore: Allow the use of plugin panels

Rename ExplorePanel to CustomContainer

Umaaz commented:
Is there any update on if this can be accepted? Or if I need to make any changes?

Elfo404 commented:
@Umaaz sorry for the delay, we are planning to take a better look at this soon, there are some unknowns we want to discuss with the team regarding Explore but a few people are off until next week.

ifrost commented:
Thanks, @Umaaz for submitting it. It's a feature that was previously approached as well (e.g. https://github.com/grafana/grafana/pull/53600)

> Why do we need this feature?
> I needed this feature, as I am developing a datasource that stores data that doesn't fall into the categories provided by the current explore page.
> Who is this feature for? 
> ***(...) want to take advantage of the explore functionality.***

Just out of curiosity - what other Explore functionalities you will use in your use case? Have you considered writing an application plugin?

Umaaz commented:
@ifrost essentially the use case of explore would be the same as say Tempo (traces). In this case there is a specific trace view that displays the trace in a way other than a table or chart. 

I have the case that the snapshot data I want to render is difficult to render as a table or chart (as are traces). So I wanted to add a specific render (which I have via a panel plugin). 

We have considered an App plugin, and will no doubt add one for some more specific functionality that would not make sense in Explore. We would still want the ability to use plugin panels in explore however, due to the reasons stated above.

If there are changes that are needed, or anything else I am able to do. Please let me know.

ifrost commented:
Thanks @Umaaz!

> If there are changes that are needed, or anything else I am able to do. Please let me know.

This works really nice and I can see great value in it. I think there's a few things that should be addressed:

> if we need a new custom meta field it might be good to add frame.meta.preferredVisualisationPluginId ?

I think this may work better as we can still use the preferredVisualizationType as the fallback. Users may use the data source plugin but not the visualization plugin, so we should have such fallback in place.

> why is only a single frame passed to the visualization?

Explore groups data by preferredVisualizationType and the same should happen with preferredVisualisationPluginId. Each visualization plugin should be capable of handling multiple data frames. Essentially it means there will be one CustomContainer per preferredVisualisationPluginId.

> ```"title":    fmt.Sprintf("Snapshot: %v:%v", snap.Tracepoint.Path, snap.Tracepoint.LineNumber),```

We don't have custom titles for visualizations in Explore. At the moment they are just generic titles indicating type of the data and would be good to keep it consistent. It'd be great to show the visualization plugin name as the title. If you need to show some extra context you can still pass it in meta and then consume it from the visualization plugin.

This is a change that would be available to use by other developers too, so I suggest marking it as `@alpha` (e.g. preferredVisualisationPluginId property). We also need some documentation on how to use it.

Umaaz committed:
Merge branch 'main' into main

Umaaz committed:
Explore: Allow the use of plugin panels

Changed CustomContainer to take all frames for plugin id.
Add field preferredVisualisationPluginId to frame metadata.
Updated decorators to check for plugin and fallback to preferredVisualisationType if plugin cannot be found.

Umaaz committed:
Explore: Allow the use of plugin panels

Handle case where there are no custom frames

ifrost commented on public/app/features/explore/Explore.tsx:
```diff
@@ -283,6 +285,36 @@ export class Explore extends React.PureComponent<Props, ExploreState> {
+  renderCustom(width: number) {
+    const { timeZone, queryResponse, absoluteRange } = this.props;
+
+    const groupedByPlugin = queryResponse?.customFrames.reduce((acc: Record<string, DataFrame[]>, cur) => {
+      // cur.meta?.preferredVisualisationPluginId could be undefined by the API, but it won't be if we
+      // got here as it is checked in decorators.ts
+      if (!acc[`${cur.meta?.preferredVisualisationPluginId}`]) {
+        acc[`${cur.meta?.preferredVisualisationPluginId}`] = [cur];
+      } else {
+        acc[`${cur.meta?.preferredVisualisationPluginId}`].push(cur);
+      }
+      return acc;
+    }, {});
```
I think lodash's `groupBy` should give the same results and be a bit more readable:

```suggestion
    const groupedByPlugin = groupBy(queryResponse?.customFrames, 'meta.preferredVisualisationPluginId')
```

Umaaz replied:
I have changed this to use lodash.

ifrost commented on public/app/features/plugins/importPanelPlugin.ts:
```diff
@@ -28,6 +28,14 @@ export function importPanelPlugin(id: string): Promise<PanelPlugin> {
+  return !!getPanelPluginMeta(id);
+}
+
+export function getPanelPluginMeta(id: string): PanelPluginMeta {
```
The same code is used above to get meta in `importPanelPlugin` so maybe we could reuse it there?

ifrost commented on public/app/features/explore/utils/decorators.ts:
```diff
@@ -34,8 +35,13 @@ export const decorateWithFrameTypeMetadata = (data: PanelData): ExplorePanelData
+  const customFrames: DataFrame[] = [];
 
   for (const frame of data.series) {
+    if (frame.meta?.preferredVisualisationPluginId && canFindPanel(frame)) {
```
Is the check for preferredVisualizationPluginId needed here? It seems `canFindPanel` checks for it anyway:

```suggestion
    if (canFindPanel(frame)) {
```

Umaaz replied:
you are correct, I have removed the additional check.

ifrost commented on public/app/features/explore/utils/decorators.test.ts:
```diff
@@ -84,6 +84,7 @@ const createExplorePanelData = (args: Partial<ExplorePanelData>): ExplorePanelDa
     tableResult: null,
     traceFrames: [],
     nodeGraphFrames: [],
+    customFrames: [],
```
I know it's not done for all the cases but it'd be good to set some customFrames here to have and example and a test coverage.

Umaaz replied:
I have added a test, but did you want it explicitly in this method as well?

ifrost replied:
Looks good, I think 4cdbf013fd10927a7c4dbf3c48c1180abfbd5414 covers it well 👍

ifrost commented on public/app/features/explore/AddToDashboard/addToDashboard.ts:
```diff
@@ -80,6 +80,9 @@ function getPanelType(queries: DataQuery[], queryResponse: ExplorePanelData) {
     if (queryResponse.traceFrames.some(hasQueryRefId)) {
       return 'traces';
     }
+    if (queryResponse.customFrames.some(hasQueryRefId)) {
```
This should also be covered in addToDashboard.test.ts

Umaaz replied:
This should now be covered by the tests.

ifrost reviewed (commented):
Thanks for addressing all the feedback @Umaaz 🙌 I've left a few more small comments to clean in up a bit. Seems to work great!

@grafana/dashboards-squad / @grafana/grafana-frontend-platform: are you okay with changes in grafana-data?

Umaaz committed:
Explore: Allow the use of plugin panels

Add test cases

ifrost commented:
LGTM 👍  Please have a look at the conflicts (hopefully should be easy to resolve), I'll run the remaining PR checks and it should be good to go!

Umaaz committed:
Merge remote-tracking branch 'origin/main' into fork-main

Umaaz commented:
Awesome @ifrost. Thanks! I have (hopefully) resolved the conflicts.

Don't forget the Go sdk will need updated for this to work from Go plugins. https://github.com/grafana/grafana-plugin-sdk-go/pull/705

ifrost reviewed (approved):
![](https://media.giphy.com/media/2wX0gLjH30bzzxpP8K/giphy.gif)
